File: api/src/routes/collections.js
import { Router } from 'express';
import {
  createCollection,
  getCollection,
  listCollections,
  updateCollection,
  deleteCollection,
  addBusinessesToCollection,
  removeBusinessFromCollection,
  listCollectionBusinesses,
} from '../db/collections.js';

const router = Router();

/**
 * @openapi
 * /api/collections:
 *   get:
 *     summary: List all collections
 *     description: Returns all collections with aggregated lead counts (totalLeads, noWebsite, newLeads). Ordered by most recently updated first.
 *     tags: [Collections]
 *     responses:
 *       200:
 *         description: Array of collections
 *         content:
 *           application/json:
 *             schema:
 *               type: array
 *               items:
 *                 $ref: '#/components/schemas/Collection'
 *   post:
 *     summary: Create a new collection
 *     description: Create a named collection with optional target location and business category tags.
 *     tags: [Collections]
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             $ref: '#/components/schemas/CollectionCreate'
 *     responses:
 *       201:
 *         description: Created collection
 *         content:
 *           application/json:
 *             schema:
 *               $ref: '#/components/schemas/Collection'
 *       400:
 *         description: name is required
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 error:
 *                   type: string
 *
 * /api/collections/{id}:
 *   get:
 *     summary: Get a collection by ID
 *     description: Returns a single collection with aggregated lead counts.
 *     tags: [Collections]
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: integer
 *         description: Collection auto-increment ID
 *     responses:
 *       200:
 *         description: Collection details
 *         content:
 *           application/json:
 *             schema:
 *               $ref: '#/components/schemas/Collection'
 *       404:
 *         description: Collection not found
 *   patch:
 *     summary: Update a collection
 *     description: Update name, location, or tags of a collection.
 *     tags: [Collections]
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: integer
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             $ref: '#/components/schemas/CollectionUpdate'
 *     responses:
 *       200:
 *         description: Updated collection
 *         content:
 *           application/json:
 *             schema:
 *               $ref: '#/components/schemas/Collection'
 *       404:
 *         description: Collection not found
 *   delete:
 *     summary: Delete a collection
 *     description: Permanently deletes a collection and its business associations (CASCADE).
 *     tags: [Collections]
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: integer
 *     responses:
 *       204:
 *         description: Deleted successfully (no content)
 *       404:
 *         description: Collection not found
 *
 * /api/collections/{id}/businesses:
 *   get:
 *     summary: List businesses in a collection
 *     description: Paginated list of result objects associated with this collection, ordered by most recently added.
 *     tags: [Collections]
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: integer
 *       - in: query
 *         name: page
 *         schema:
 *           type: integer
 *           default: 1
 *         description: Page number (1-based)
 *       - in: query
 *         name: limit
 *         schema:
 *           type: integer
 *           default: 20
 *         description: Results per page (max 100)
 *     responses:
 *       200:
 *         description: Paginated list of results
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 rows:
 *                   type: array
 *                   items:
 *                     $ref: '#/components/schemas/Result'
 *                 total:
 *                   type: integer
 *                 page:
 *                   type: integer
 *                 limit:
 *                   type: integer
 *       404:
 *         description: Collection not found
 *   post:
 *     summary: Add businesses to a collection
 *     description: Associate one or more result IDs with this collection. Duplicates are silently ignored.
 *     tags: [Collections]
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: integer
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             $ref: '#/components/schemas/AddBusinessesRequest'
 *     responses:
 *       200:
 *         description: Businesses added
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 success:
 *                   type: boolean
 *       400:
 *         description: resultIds array is required or empty
 *       404:
 *         description: Collection not found
 *
 * /api/collections/{id}/businesses/{resultId}:
 *   delete:
 *     summary: Remove a business from a collection
 *     description: Remove a single result association from the collection. Does not delete the result itself.
 *     tags: [Collections]
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         schema:
 *           type: integer
 *         description: Collection ID
 *       - in: path
 *         name: resultId
 *         required: true
 *         schema:
 *           type: string
 *         description: Result UUID to remove
 *     responses:
 *       204:
 *         description: Removed successfully (no content)
 *       404:
 *         description: Collection not found
 */

router.get('/', (req, res) => {
  try {
    const collections = listCollections();
    res.json(collections);
  } catch (err) {
    console.error('GET /api/collections error:', err);
    res.status(500).json({ error: err.message });
  }
});

router.post('/', (req, res) => {
  try {
    const { name, location, tags } = req.body;
    if (!name || !name.trim()) {
      return res.status(400).json({ error: 'name is required' });
    }
    const collection = createCollection({ name: name.trim(), location: location || null, tags: tags || null });
    res.status(201).json(collection);
  } catch (err) {
    console.error('POST /api/collections error:', err);
    res.status(500).json({ error: err.message });
  }
});

router.get('/:id', (req, res) => {
  try {
    const collection = getCollection(parseInt(req.params.id));
    if (!collection) return res.status(404).json({ error: 'Collection not found' });
    res.json(collection);
  } catch (err) {
    console.error('GET /api/collections/:id error:', err);
    res.status(500).json({ error: err.message });
  }
});

router.patch('/:id', (req, res) => {
  try {
    const fields = {};
    for (const key of ['name', 'location', 'tags']) {
      if (req.body[key] !== undefined) fields[key] = req.body[key];
    }
    const collection = updateCollection(parseInt(req.params.id), fields);
    if (!collection) return res.status(404).json({ error: 'Collection not found' });
    res.json(collection);
  } catch (err) {
    console.error('PATCH /api/collections/:id error:', err);
    res.status(500).json({ error: err.message });
  }
});

router.delete('/:id', (req, res) => {
  try {
    const collection = getCollection(parseInt(req.params.id));
    if (!collection) return res.status(404).json({ error: 'Collection not found' });
    deleteCollection(parseInt(req.params.id));
    res.status(204).end();
  } catch (err) {
    console.error('DELETE /api/collections/:id error:', err);
    res.status(500).json({ error: err.message });
  }
});

router.get('/:id/businesses', (req, res) => {
  try {
    const collection = getCollection(parseInt(req.params.id));
    if (!collection) return res.status(404).json({ error: 'Collection not found' });

    const page = Math.max(1, parseInt(req.query.page) || 1);
    const limit = Math.min(100, Math.max(1, parseInt(req.query.limit) || 20));
    const result = listCollectionBusinesses(parseInt(req.params.id), { page, limit });
    res.json(result);
  } catch (err) {
    console.error('GET /api/collections/:id/businesses error:', err);
    res.status(500).json({ error: err.message });
  }
});

router.post('/:id/businesses', (req, res) => {
  try {
    const collection = getCollection(parseInt(req.params.id));
    if (!collection) return res.status(404).json({ error: 'Collection not found' });

    const { resultIds } = req.body;
    if (!resultIds || !Array.isArray(resultIds) || resultIds.length === 0) {
      return res.status(400).json({ error: 'resultIds array is required' });
    }

    addBusinessesToCollection(parseInt(req.params.id), resultIds);
    res.json({ success: true });
  } catch (err) {
    console.error('POST /api/collections/:id/businesses error:', err);
    res.status(500).json({ error: err.message });
  }
});

router.delete('/:id/businesses/:resultId', (req, res) => {
  try {
    const collection = getCollection(parseInt(req.params.id));
    if (!collection) return res.status(404).json({ error: 'Collection not found' });

    removeBusinessFromCollection(parseInt(req.params.id), req.params.resultId);
    res.status(204).end();
  } catch (err) {
    console.error('DELETE /api/collections/:id/businesses/:resultId error:', err);
    res.status(500).json({ error: err.message });
  }
});

export default router;

File: api-python/routes/collections.py
import csv
import io
import json
import uuid

from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import StreamingResponse

from database import (
    create_collection,
    get_collection,
    list_collections,
    update_collection,
    delete_collection,
    add_businesses_to_collection,
    remove_business_from_collection,
    list_collection_businesses,
    get_db,
)
from models import (
    CollectionCreate,
    CollectionUpdate,
    CollectionResponse,
    AddBusinessesRequest,
)

router = APIRouter(prefix='/api/collections', tags=['Collections'])


@router.get('', response_model=list[CollectionResponse],
    summary='List all collections',
    description='Returns all collections with aggregated lead counts.')
async def list_collections_endpoint():
    return list_collections()


@router.post('', response_model=CollectionResponse, status_code=201,
    summary='Create a new collection',
    description='Create a named collection with optional location and tags.')
async def create_collection_endpoint(body: CollectionCreate):
    if not body.name or not body.name.strip():
        raise HTTPException(400, 'name is required')
    return create_collection(name=body.name.strip(), location=body.location, tags=body.tags)


@router.get('/{collection_id}', response_model=CollectionResponse,
    summary='Get a collection by ID',
    description='Returns collection details with aggregated lead counts.')
async def get_collection_endpoint(collection_id: int):
    collection = get_collection(collection_id)
    if not collection:
        raise HTTPException(404, 'Collection not found')
    return collection


@router.patch('/{collection_id}', response_model=CollectionResponse,
    summary='Update a collection',
    description='Update name, location, tags, or status of a collection.')
async def update_collection_endpoint(collection_id: int, body: CollectionUpdate):
    collection = get_collection(collection_id)
    if not collection:
        raise HTTPException(404, 'Collection not found')
    fields = {k: v for k, v in body.model_dump(exclude_unset=True).items() if v is not None}
    return update_collection(collection_id, **fields)


@router.delete('/{collection_id}', status_code=204,
    summary='Delete a collection',
    description='Permanently deletes a collection and its business associations.')
async def delete_collection_endpoint(collection_id: int):
    collection = get_collection(collection_id)
    if not collection:
        raise HTTPException(404, 'Collection not found')
    delete_collection(collection_id)


@router.get('/{collection_id}/export/ghl-csv',
    summary='Export collection as GoHighLevel CSV',
    description='Downloads a GoHighLevel-compatible CSV of all contacts in this collection.')
async def export_ghl_csv_endpoint(collection_id: int):
    collection = get_collection(collection_id)
    if not collection:
        raise HTTPException(404, 'Collection not found')

    db = get_db()
    rows = db.execute(
        '''SELECT r.*
           FROM collection_businesses cb
           JOIN results r ON r.id = cb.result_id
           WHERE cb.collection_id = ?
           ORDER BY cb.added_at DESC''',
        (collection_id,),
    ).fetchall()

    if not rows:
        raise HTTPException(404, 'No businesses in this collection')

    output = io.StringIO()
    writer = csv.writer(output)

    writer.writerow([
        'First Name', 'Last Name', 'Email', 'Phone', 'Company Name',
        'Address', 'Website', 'Tags', 'Notes', 'Source',
    ])

    collection_tags = collection.get('tags') or []

    for row in rows:
        row = dict(row)
        name = row.get('name') or ''
        address = row.get('address') or ''
        phone = row.get('phone') or ''
        website = row.get('website') or row.get('network_site') or ''
        notes = row.get('notes') or ''
        categories = row.get('categories')
        if isinstance(categories, str):
            categories = json.loads(categories)

        tags = list(collection_tags)
        if categories:
            tags.extend(categories)
        tags_str = ', '.join(tags)

        first_name = ''
        last_name = ''
        email = ''
        
        if not first_name and not email and not phone:
            first_name = f'Temporary {uuid.uuid4()}'

        writer.writerow([
            first_name,
            last_name,
            email,
            phone,
            name,
            address,
            website,
            tags_str,
            notes,
            'FindBusinesses',
        ])

    output.seek(0)
    filename = f"{collection['name'].replace(' ', '_')}_ghl_import.csv"
    return StreamingResponse(
        iter([output.getvalue()]),
        media_type='text/csv',
        headers={'Content-Disposition': f'attachment; filename="{filename}"'},
    )


@router.get('/{collection_id}/businesses',
    summary='List businesses in a collection',
    description='Paginated list of results associated with this collection.')
async def list_businesses_endpoint(
    collection_id: int,
    page: int = Query(1, ge=1, description='Page number'),
    limit: int = Query(20, ge=1, le=100, description='Results per page (max 100)'),
):
    collection = get_collection(collection_id)
    if not collection:
        raise HTTPException(404, 'Collection not found')
    return list_collection_businesses(collection_id, page=page, limit=limit)


@router.post('/{collection_id}/businesses',
    summary='Add businesses to a collection',
    description='Associate one or more result IDs with this collection.')
async def add_businesses_endpoint(collection_id: int, body: AddBusinessesRequest):
    collection = get_collection(collection_id)
    if not collection:
        raise HTTPException(404, 'Collection not found')
    if not body.resultIds:
        raise HTTPException(400, 'resultIds array is required')
    add_businesses_to_collection(collection_id, body.resultIds)
    return {'success': True}


@router.delete('/{collection_id}/businesses/{result_id}', status_code=204,
    summary='Remove a business from a collection',
    description='Remove a single result from the collection.')
async def remove_business_endpoint(collection_id: int, result_id: str):
    collection = get_collection(collection_id)
    if not collection:
        raise HTTPException(404, 'Collection not found')
    remove_business_from_collection(collection_id, result_id)
